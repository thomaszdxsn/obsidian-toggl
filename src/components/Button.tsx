import { css } from '@emotion/css'
import clsx from 'clsx'
import React from 'react'
import { ClipLoader } from 'react-spinners'

interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode
  isLoading?: boolean
  loadingIconSize?: number
}


export const Button = ({ children, isLoading, loadingIconSize = 13, className, ...props }: Props) => {
  return <button
    {...props}
    disabled={isLoading || props.disabled}
    className={
      clsx(css`
        cursor: ${isLoading ? "wait" : "pointer"};
      `, className)
    }
  >{isLoading ?
    <ClipLoader size={loadingIconSize}
      data-testid="loading-icon"
      cssOverride={{
        borderColor: "var(--accent-h)",
      }}
    />
    : children}</button>
}