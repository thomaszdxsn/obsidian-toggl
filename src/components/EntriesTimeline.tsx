import React, { useMemo } from "react";
import { Project, TimeEntry } from "../interfaces";
import { Timeline } from "./Timeline";
import { calcPercentage, formatSeconds, formatTime } from "../utils";
import { css } from "@emotion/css";
import clsx from "clsx";
import { TimeDisplay } from "./TimeDisplay";

interface Props {
  className?: string;
  entries: TimeEntry[];
  projectDict: Record<string, Project>;
  entryMinHeight?: number;
  equalEntryHeight?: boolean;
}

export const EntriesTimeline = ({
  className,
  entries,
  projectDict,
  entryMinHeight = 100,
  equalEntryHeight = true,
}: Props) => {
  const items = useMemo(() => {
    const totalDuration = entries.reduce((acc, entry) => {
      return acc + entry.duration;
    }, 0);
    return entries
      .slice()
      .sort((a, b) => (a.start < b.start ? 1 : -1))
      .map((entry) => {
        const project = entry.project_id
          ? projectDict[entry.project_id]
          : undefined;
        const percentage = equalEntryHeight
          ? calcPercentage(1, entries.length)
          : calcPercentage(entry.duration, totalDuration);
        const label = (
          <div
            className={css`
              display: flex;
              justify-content: space-between;
              height: 100%;
              gap: 8px;
              padding-right: 16px;
            `}
          >
            <div
              className={css`
                flex-shrink: 0;
                display: flex;
                height: 100%;
                flex-direction: column;
                justify-content: space-between;
                min-width: 0;
                padding: 16px 0;
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
              `}
            >
              <div style={{ color: project?.color }}>{project?.name}</div>
              <div
                className={css`
                  white-space: nowrap;
                  overflow: hidden;
                  text-overflow: ellipsis;
                `}
              >
                {entry.description}
              </div>
              <div
                className={css`
                  color: var(--text-accent);
                  white-space: nowrap;
                  overflow: hidden;
                  text-overflow: ellipsis;
                `}
              >
                üè∑Ô∏è {entry.tags.join(",")}
              </div>
            </div>
            <div
              className={css`
                height: 100%;
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: end;
                gap: 8px;
                flex-shrink: 0;
              `}
            >
              <TimeDisplay
                className={css`
                  font-size: 1rem;
                  font-weight: var(--font-semibold);
                `}
              >
                {formatSeconds(entry.duration)}
              </TimeDisplay>
              <TimeDisplay
                className={css`
                  font-size: 0.75rem;
                `}
              >
                {formatTime(entry.start)} - {formatTime(entry.stop)}
              </TimeDisplay>
            </div>
          </div>
        );
        return {
          color: project?.color || "var(--background-primary)",
          label: label,
          percentage: percentage,
        };
      });
  }, [entries, projectDict]);
  const totalHeight = useMemo(() => {
    const smallestPercentage = Math.min(
      ...items.map((item) => item.percentage),
    );
    return items.reduce((acc, item) => {
      return acc + (item.percentage / smallestPercentage) * entryMinHeight;
    }, 0);
  }, [items, entryMinHeight]);
  return (
    <Timeline
      direction="vertical"
      className={clsx(
        className,
        css`
          height: ${totalHeight}px;
        `,
      )}
      items={items}
      size="medium"
      gap={1}
      finished
    />
  );
};
