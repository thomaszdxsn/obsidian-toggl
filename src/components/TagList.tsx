import React from 'react'
import {  Tag } from "../interfaces"
import { css } from '@emotion/css'

interface Props {
  data: Tag[]
}

export const TagList = ({ data }: Props) => {
  return (
    <div>
      {data.map((tag) => (
        <div
          key={tag.id}
          className={css`
              padding: 0.5rem 0;
              border-bottom: 1px solid var(--color-border, gray);
             `}
        >
          🏷️ {tag.name}
        </div>
      ))}
    </div>
  )
}