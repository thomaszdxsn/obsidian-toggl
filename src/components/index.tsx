export { CurrentTimer } from './CurrentTimer'
export { MainView } from './MainView'
export { Provider } from './Provider'
export { TimerCard } from './TimerCard'
export { TimerList } from './TimerList'
export { TimerForm } from './TimerForm'
export { Button } from './Button'
export { Timeline } from './Timeline'