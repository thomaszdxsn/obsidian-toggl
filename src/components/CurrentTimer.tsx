import React from "react";
import { useAtom, useAtomValue, useSetAtom } from "jotai";
import { useStopTimerMutation } from "../hooks";
import {
  currentEntryAtom,
  currentEntryProjectAtom,
  currentTimerAtom,
  passedSecondsAtom,
  passedTimeAtom,
} from "../atoms";
import { css } from "@emotion/css";
import { FiPause } from "react-icons/fi";
import { TimeDisplay } from "./TimeDisplay";
import { Button } from "./Button";
import { generateTimerId } from "../utils";

export const CurrentTimer = ({
  placeholder = "No time entry running",
}: {
  placeholder?: string;
}) => {
  const passed = useAtomValue(passedTimeAtom);
  const setPassedNumbers = useSetAtom(passedSecondsAtom);
  const [currentEntry, setCurrentEntry] = useAtom(currentEntryAtom);
  const currentTimer = useAtomValue(currentTimerAtom);
  const stopTimerMutation = useStopTimerMutation(
    currentTimer ? generateTimerId(currentTimer) : undefined,
  );
  const currentProject = useAtomValue(currentEntryProjectAtom);
  const onStop = () => {
    if (!currentEntry) {
      return;
    }
    stopTimerMutation.mutate(
      { timeEntryId: currentEntry.id, workspaceId: currentEntry.workspace_id },
      {
        onSuccess: () => {
          setCurrentEntry(null);
          setPassedNumbers(null);
        },
      },
    );
  };
  if (!passed) {
    return (
      <div
        className={css`
          height: 100px;
          display: flex;
          justify-content: center;
          align-items: center;
          border-radius: var(--radius-m);
          background-color: var(--background-primary);
        `}
      >
        {placeholder}
      </div>
    );
  }
  return (
    <div
      className={css`
        display: flex;
        flex-direction: column;
        background-color: var(--background-primary);
        padding: var(--size-4-4);
        border-radius: var(--radius-m);
        height: 100px;
        container-type: inline-size;
      `}
    >
      <Button
        onClick={onStop}
        isLoading={stopTimerMutation.isPending}
        className={css`
          flex: 1;
          cursor: ${stopTimerMutation.isPending ? "wait" : "pointer"};
          margin-bottom: var(--size-4-2);
        `}
      >
        <FiPause />
      </Button>
      <div
        className={css`
          display: flex;
          justify-content: center;
          align-items: center;
          flex-wrap: wrap;
          @container (min-width: 200px) {
            justify-content: space-between;
          }
        `}
      >
        <div>
          {currentProject && (
            <div style={{ color: currentProject.color }}>
              {currentProject.name}
            </div>
          )}
          {currentEntry && (
            <div
              className={css`
                font-size: var(--font-smallest);
                color: var(--text-muted);
              `}
            >
              {currentEntry.description}
            </div>
          )}
        </div>
        <TimeDisplay
          className={css`
            font-size: var(--font-ui-large);
            font-weight: var(--font-semibold);
          `}
        >
          {passed}
        </TimeDisplay>
      </div>
    </div>
  );
};
