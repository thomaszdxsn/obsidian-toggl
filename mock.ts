import { faker } from '@faker-js/faker'
import { Project, TimeEntry } from 'src/interfaces'


export function createRandomTimeEntry(): TimeEntry {
  return {
    id: faker.number.int(),
    pid: faker.number.int(),
    project_id: faker.number.int(),
    description: faker.lorem.sentence(),
    duronly: faker.datatype.boolean(),
    wid: faker.number.int(),
    tags: [faker.lorem.word()],
    tag_ids: [faker.number.int()],
    start: faker.date.recent().toISOString(),
    stop: faker.date.recent().toISOString(),
    duration: faker.number.int(),
    billable: faker.datatype.boolean(),
    at: faker.date.recent().toISOString(),
    server_deleted_at: faker.date.recent().toISOString(),
    user_id: faker.number.int(),
    task_id: faker.number.int(),
    uid: faker.number.int(),
    tid: faker.number.int(),
    workspace_id: faker.number.int(),
  }
}

export function createRandomProject(): Project {
  return {
    id: faker.number.int(),
    wid: faker.number.int(),
    cid: faker.number.int(),
    name: faker.lorem.word(),
    billable: faker.datatype.boolean(),
    is_private: faker.datatype.boolean(),
    active: faker.datatype.boolean(),
    at: faker.date.recent().toISOString(),
    created_at: faker.date.recent().toISOString(),
    color: faker.internet.color(),
    auto_estimates: faker.datatype.boolean(),
    actual_hours: faker.number.int(),
    rate: faker.number.float(),
    template: faker.datatype.boolean(),
    template_id: faker.number.int(),
    actual_seconds: faker.number.int(),
    client_id: faker.number.int(),
    currency: faker.finance.currencyCode(),
    current_period: {
      at: faker.date.recent().toISOString(),
    },
    end_date: faker.date.recent().toISOString(),
    estimated_hours: faker.number.int(),
    estimated_seconds: faker.number.int(),
    fixed_fee: faker.number.float(),
    rate_last_updated: faker.date.recent().toISOString(),
    recurring: faker.datatype.boolean(),
    start_date: faker.date.recent().toISOString(),
    server_deleted_at: faker.date.recent().toISOString(),
    status: faker.lorem.word(),
    workspace_id: faker.number.int(),
  }
}
