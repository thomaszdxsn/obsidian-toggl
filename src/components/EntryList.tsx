import React from "react";
import { useAtomValue } from "jotai";
import { projectDictAtom, todayTimeEntriesAtom } from "src/atoms";
import { EntriesTimeline } from "./EntriesTimeline";
import { isActiveEntry } from "src/utils";
import { css } from "@emotion/css";

export const EntryList = () => {
  const todayEntries = useAtomValue(todayTimeEntriesAtom);
  const entries = todayEntries.filter((entry) => !isActiveEntry(entry));
  const projectDict = useAtomValue(projectDictAtom);
  console.log({ entries });
  return (
    <div
      className={css`
        max-height: 30vh;
        overflow-y: auto;
      `}
    >
      <EntriesTimeline entries={entries} projectDict={projectDict} />
    </div>
  );
};
