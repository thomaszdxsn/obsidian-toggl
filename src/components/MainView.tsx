import { css } from "@emotion/css"
import React, { useMemo } from "react"
import { usePlugin } from "../hooks"
import { ProjectModal, TimerDetailModal } from "../plugin/modal"
import { CurrentTimer } from "./CurrentTimer"
import { TimerList } from "./TimerList"
import { FiCornerUpLeft, FiFolder, FiPlus } from 'react-icons/fi'
import { Button } from "./Button"
import { useAtomValue, useSetAtom } from "jotai"
import { activeProjectsAtom, projectDictAtom, todayTimeEntriesAtom, viewAtom } from "src/atoms"
import dayjs from "dayjs"
import { Timeline } from "./Timeline"
import { RefreshButton } from "./RefreshButton"
import { EntryList } from "./EntryList"
import { ProjectList } from "./ProjectList"

export const MainView = () => {
  const currentView = useAtomValue(viewAtom)
  const viewContent = useMemo(() => {
    switch (currentView) {
      case "homepage":
        return <HomePageView />
      case "projects":
        return <ProjectsView />
      case "tags":
        return <TagsView />
      default:
        return null
    }
  }, [currentView])
  return (
    <div
      className={css`
				display: flex;
				flex-direction: column;
				gap: var(--size-4-2);
			`}
    >
      <TodayTimeline />
      <CurrentTimer />
      <section className={css`
				display: flex;
				flex-direction: column;
				gap: var(--size-4-2);
			`}>
        {viewContent}
      </section>
    </div>
  )
}


export const TodayTimeline = () => {
  const entries = useAtomValue(todayTimeEntriesAtom)
  const projectDict = useAtomValue(projectDictAtom)
  const timelineItems = useMemo(() => {
    const total = entries.reduce((acc, entry) => {
      const duration = entry.duration !== -1 ? entry.duration : dayjs().diff(entry.start, "second")
      return acc + duration
    }, 0)
    return entries.sort((a, b) => a.start < b.start ? - 1 : 1).map(entry => {
      const project = entry.project_id ? projectDict[entry.project_id] : null
      const duration = entry.duration !== -1 ? entry.duration : dayjs().diff(entry.start, "second")
      const percentage = duration / total
      return {
        color: project?.color || "var(--background-primary)",
        label: project?.name,
        percentage: percentage
      }
    })

  }, [entries, projectDict])
  return (
    <div className={css`
        background: var(--background-primary);
        border-start-end-radius: 1rem;
        border-end-end-radius: 1rem;
        flex: 1;
      `}>
      <Timeline items={timelineItems} direction="horizontal" showLabel={false} />
    </div>
  )
}


export const HomePageView = () => {
  const plugin = usePlugin()
  const app = plugin.app
  const switchView = useSetAtom(viewAtom)
  const onAddTimer = () => {
    const modal = new TimerDetailModal(app, plugin)
    modal.open()
  }
  const onSwitchToProjectsView = () => {
    switchView("projects")
  }
  return (
    <>
      <div className={css`
          display: flex;
          gap: var(--size-4-2);
          justify-content: flex-end;
        `}>
        <Button onClick={onSwitchToProjectsView} title="projects">
          <FiFolder />
        </Button>
        <RefreshButton />
        <Button onClick={onAddTimer} title="add new timer">
          <FiPlus />
        </Button>
      </div>
      <TimerList plugin={plugin} />
      <EntryList />
    </>
  )
}


export const ProjectsView = () => {
  const plugin = usePlugin()
  const switchView = useSetAtom(viewAtom)
  const backToHomePage = () => switchView("homepage")
  const onAddProject = () => {
    const modal = new ProjectModal(plugin)
    modal.open()
  }
  const projects = useAtomValue(activeProjectsAtom)
  return (
    <>
      <div className={css`
      display: flex;
      justify-content: space-between;
    `}>
        <Button onClick={backToHomePage} title="projects">
          <FiCornerUpLeft />
        </Button>
        <div className={css`
          display: flex;
          gap: var(--size-4-2);
        `}>
          <Button onClick={onAddProject} title="add new projeect">
            <FiPlus />
          </Button>
        </div>
      </div>
      <ProjectList data={projects} />
    </>
  )
}


export const TagsView = () => {
  return null
}