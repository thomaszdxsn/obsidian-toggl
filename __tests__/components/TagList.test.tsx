import { TagList } from "../../src/components/TagList";
import "@testing-library/jest-dom";
import React from 'react';
import { render } from '@testing-library/react';
import { createRandomTag } from "../../mock";


it("render correctly", () => {
  const tags = Array.from({length: 10}, (_, i) => createRandomTag({
    name: `Tag Name ${i}`
  }))
  const {getAllByText, container} = render(<TagList data={tags} />)
  const tagNames = getAllByText(/Tag Name/)
  expect(tagNames).toHaveLength(10)
  expect(container).toMatchSnapshot()
})