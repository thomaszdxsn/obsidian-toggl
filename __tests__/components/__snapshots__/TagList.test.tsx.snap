// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div>
  <div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 0
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 1
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 2
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 3
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 4
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 5
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 6
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 7
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 8
    </div>
    <div
      class="css-kj0z6i"
    >
      🏷️ 
      Tag Name 9
    </div>
  </div>
</div>
`;
